File: ruby_compiler/lex/analyser.rb
module Lex
  
  class Analyser
    
    def initialize(content)
      @content = content
      @fixed_content = @content + ' '
      @t = Transducer.new
      @line = 1
      @column = 0
    end
    
    def get_next_token
      token = Token.new
      type = Token::Type::IGNORED
      value = nil
      line_end = @line
      column = @column

      if @content.empty?
        token.type = Token::Type::END_OF_FILE
        return token
      end
      
      # Read ignored chars
      while @content.size > 0 and type == Token::Type::IGNORED
        # do nothing, chars are being ignored
        type = @t.consume_input(@fixed_content[0], @fixed_content[1])
        value = @content[0,1]
        @content = @content[1,@content.size]
        @fixed_content = @fixed_content[1,@fixed_content.size]
        
        @column += 1
        if value[0] == ?\n
          line_end += 1
          @column = 0
        end
      end

      @line = line_end
      column = @column
      
      # Read token
      while @content.size > 0 and type == Token::Type::INCOMPLETE
        type = @t.consume_input(@fixed_content[0], @fixed_content[1])
        value += @content[0,1]
        @content = @content[1,@content.size]
        @fixed_content = @fixed_content[1,@fixed_content.size]
        
        @column += 1
        if value[-1] == ?\n
          line_end += 1
          @column = 0
        end
      end
      
      value = nil if type == Token::Type::IGNORED
      
      if type == Token::Type::INCOMPLETE
        type = Token::Type::INVALID
      elsif type == Token::Type::IDENTIFIER and Semantic::TableReservedWords.include?(value)
        type = Token::Type::RESERVED_WORD
      end
      
      token.value = value
      token.index = table_value(value, type)
      token.type = type
      token.line = @line
      token.column = column
      
      @line = line_end
      
      return token
    end
    
    def table_value(value, type)
      index = -1
      
      if type == Token::Type::INT_NUMBER
        index = Semantic::TableNumbers.add(value)
      elsif type == Token::Type::RESERVED_WORD
        index = Semantic::TableReservedWords.index(value)
      elsif type == Token::Type::IDENTIFIER
        index = Semantic::TableSymbols.add(value)
      elsif type == Token::Type::SPECIAL
        index = Semantic::TableSpecial.index(value)
      end
      
      return index
    end
    
  end
  
end

File: compiler/analyser.c
/*
 *  analyser.c
 *  compiler
 *
 *  Created by Filipe Morgado Simões de Campos e Rafael Barbolo Lopes on 24/09/10.
 *  Copyright 2010 __MyCompanyName__. All rights reserved.
 *
 */

#include "analyser.h"


token_type get_next_token(){

	//char a_char;
	int i = 0;
	
	init_reader(read_head);
	
	new_token(token);
//	new_transducer(transducer); // Vai precisar disso mesmo???        !!!!!       liberar essa linha depois q o transducer tiver pronto   !!!!!!!!!!!!!!!
	
	
	
	read_head = get_next_char("../resources/test_lex.poli", read_head);
	token.column++;
	
	// If the file is empty, return a empty type token.
	if (read_head.current == EOF) {
		token.type = TOKEN_TYPE_END_OF_FILE;
		return token;
	}
	
	
	// Read ignored chars
	while (read_head.current != EOF && token.type == TOKEN_TYPE_IGNORED) {
		// do nothing, chars are being ignored
		
//		token.type = transducer.consume_input(read_head.current, read_head.next);        !!!!!       liberar essa linha depois q o transducer tiver pronto
		token.value[0] = read_head.current;
		
		read_head = get_next_char("../resources/test_lex.poli", read_head);
		token.column++;
		
		if (token.value[0] == '\n') {
			token.line++;
			token.column = 0;
		}
	}
	
	
	// Read token
	while (read_head.current != EOF && token.type == TOKEN_TYPE_INCOMPLETE) {
		
//		token.type = transducer.consume_input(read_head.current, read_head.next);         !!!!!      liberar essa linha depois q o transducer tiver pronto
		token.value[i] = read_head.current;
		
		read_head = get_next_char("../resources/test_lex.poli", read_head);
		token.column++;
		
		// ????????  if value[-1] == ?\n  ???????????? o que é esse [-1]?
		if (token.value[i] == '\n') {
			token.line++;
			token.column = 0;
		}		
		
		i++;
	}
	
	
	
	if (token.type == TOKEN_TYPE_IGNORED) {
		for (int i = 0 ; i < 100; i++) {
			token.value[i] = 0;
		}
	}
	
	if (token.type == TOKEN_TYPE_INCOMPLETE) {
		token.type = TOKEN_TYPE_INVALID; 
	}
	else {
//		if (token.type == TOKEN_TYPE_IDENTIFIER &&      !!!!!     Semantic::TableReservedWords.include?(value)      !!!!!!!       ) {
//			token.type = TOKEN_TYPE_RESERVED_WORD;
//		}
	}

	
	// !!!!!!!!!!!!!!!!!!!!!!
	//token.index = table_value(value, type)       !!!!!!        liberar essa linha e fazer a função table_value depois q as tabelas ficarem prontas
	// !!!!!!!!!!!!!!!!!!!!!!	

	return token;
}

///*

